scipy/stats/_stats_py.py

```diff
@@ -1007,11 +1007,17 @@ def moment(a, moment=1, axis=0, nan_policy='propagate', *, center=None):
     """
     a, axis = _chk_asarray(a, axis)
 
-    if center is None:
-        center = a.mean(axis, keepdims=True)
     # for array_like moment input, return a value for each.
     if not np.isscalar(moment):
-        mmnt = [_moment(a, i, axis, mean=center) for i in moment]
+        # Calculated the mean once at most, and only if it will be used
+        calculate_mean = center is None and np.any(np.asarray(moment) > 1)
+        mean = a.mean(axis, keepdims=True) if calculate_mean else None
+        mmnt = []
+        for i in moment:
+            if center is None and i > 1:
```

Comment:
This makes much sense, however I detected some duplicate code here. I pushed something which you may accept.

Reply:
I don't see anything. Let's discuss the duplicate code you see before doing anything.

Reply:
> It doesn't look like you pushed anything? Let's discuss the duplicate code you see before doing anything.

I don't have permissions to push to your `gh18152` branch. I think my changes explain the duplicate code best, I can revert the changes afterwards of course. In short, `center is None and np.any(np.asarray(moment) > 1)` is checked both when assigning the boolean `calculate_mean`, and when iterating the moments as well.

It's not _exactly_ duplicate, since `calculate_mean` is `True` even if only 1 moment is `<=1`, but I still think that my changes makes the code clearer and cleaner.

Reply:
```
commit 42c2502b97ca156cf6291fc3baee0095f51b1722
Author: Doron Behar <doron.behar@gmail.com>
Date:   Sun Apr 9 10:39:33 2023 +0300

    MAINT: stats.moment: don't check center is None

diff --git a/scipy/stats/_stats_py.py b/scipy/stats/_stats_py.py
index 7be7a0669..1ec589f1e 100644
--- a/scipy/stats/_stats_py.py
+++ b/scipy/stats/_stats_py.py
@@ -1009,12 +1009,10 @@ def moment(a, moment=1, axis=0, nan_policy='propagate', *, center=None):
 
     # for array_like moment input, return a value for each.
     if not np.isscalar(moment):
-        # Calculated the mean once at most, and only if it will be used
-        calculate_mean = center is None and np.any(np.asarray(moment) > 1)
-        mean = a.mean(axis, keepdims=True) if calculate_mean else None
         mmnt = []
         for i in moment:
             if center is None and i > 1:
+                mean = a.mean(axis, keepdims=True)
                 mmnt.append(_moment(a, i, axis, mean=mean))
             else:
                 mmnt.append(_moment(a, i, axis, mean=center))
```

Reply:
I considered that, of course, and I thought it was simpler, too.
I decided against because it is far more efficient to add a calculation of 
```python3
calculate_mean = center is None and np.any(np.asarray(moment) > 1)
```
before the `for` loop than to calculate 
```python3
a.mean(axis, keepdims=True)` 
```
in _every_ iteration of the `for` loop.

Reply:
That makes sense :+1:.